src/instructlab/cli/model/upload.py

```diff
@@ -0,0 +1,239 @@
+            fg="green",
+        )
+
+    def upload_safetensors(self) -> None:
+        try:
+            hf_api = HfApi()
+            resp = hf_api.upload_folder(
+                folder_path=self.local_model_path,
+                repo_id=self.destination,
+                revision=self.release,
+                token=self.hf_token,
+            )
+            logger.debug(f"Hugging Face response:\n{resp}")
+        except Exception as exc:
+            click.secho(
+                f"\nUploading safetensors model at {self.local_model_path} failed with the following Hugging Face Hub error:\n{DEFAULT_INDENT}{exc}",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+        click.secho(
+            f"\nUploading safetensors model at {self.local_model_path} succeeded!",
+            fg="green",
+        )
+
+
+@click.command()
+@click.option(
+    "--model",
+    "-m",
+    required=True,
+    help="Path to the model to upload or name of an existing checkpoint.",
+)
+@click.option(
+    "--dest-type",
+    type=click.Choice(["hf", "oci", "s3"]),
+    default="hf",
+    help="The type of destination to upload to - can be 'hf', 'oci', or 's3'",
+)
+@click.option(
+    "--destination",
+    "-d",
+    type=str,
+    required=True,
+    help="Destination for the model to be uploaded to. Ex: for a Hugging Face repo, should be '<username>/<repo-name>'",
+)
+@click.option(
+    "--release",
+    "-rl",
+    default="main",
+    show_default=True,
+    help="The revision to upload the model to - e.g. a branch for Hugging Face repositories.",
+)
+@click.option(
+    "--hf-token",
+    default="",
+    envvar="HF_TOKEN",
+    help="User access token for connecting to the Hugging Face Hub.",
+)
+@click.pass_context
+@clickext.display_params
+def upload(ctx, model, dest_type, destination, release, hf_token):
+    """Uploads model to a specified location"""
+    uploader = None
+
+    if dest_type == "hf":
+        uploader = HFModelUploader(
+            ctx=ctx,
+            model=model,
+            destination=destination,
+            release=release,
+            hf_token=hf_token,
+        )
+    elif dest_type == "oci":
+        # TODO: update this when OCI uploading is supported
+        click.secho(
+            f"Uploading of type {dest_type} is not yet supported",
+            fg="yellow",
+        )
+        raise click.exceptions.Exit(1)
+    elif dest_type == "s3":
+        # TODO: update this when S3 uploading is supported
+        click.secho(
+            f"Uploading of type {dest_type} is not yet supported",
+            fg="yellow",
+        )
+        raise click.exceptions.Exit(1)
+    else:
+        click.secho(
+            f"{dest_type} matches neither Hugging Face, OCI registry, or an S3-compatable format.\nPlease supply a supported dest_type",
+            fg="red",
+        )
+        raise click.exceptions.Exit(1)
+
+    try:
+        uploader.upload()
```

Comment:
generally speaking, with the client server re-architecutre, we are trying to keep "click" functionality in `src/instructlab/cli` and actual operations in the `src/instructlab/model` folder.

I think we should enforce this since me, @alinaryan, and @jaideepr97 are in the process of re-architecting all of ilab to follow this format. That being said, I'll leave it up to you what you think is just click processing and flag reading, and what is actual operational code. I encourage  moving the operational code to `src/instructlab/model/upload.py` in order for it to be considered for Api implemetations, process mgmt, etc.

Reply:
I was explicitly told to do this last week - given that as well as the PR has been through multiple review cycles already I would prefer not to make such a large change this late in.

Reply:
I will admit I didn't review immediately, sorry about that! I am curious to know where this feedback was received to explicitly do it this way?

Reply:
It was actually a conversation we had - originally I was putting things in the `model` directory and having issues, and you told me to put things in the `cli` directory instead. Maybe there was a crossed wire there, but since I hadn't heard anything from @jaideepr97 and @alinaryan (who have already reviewed this PR) I assumed it was fine.

Let's talk more about this offline.

Reply:
I agree with @cdoern - I had only done a brief initial pass so far and was planning to look through again but yeah, I guess I could've mentioned that earlier

would be ideal to follow the split structure if we are introducing a new command

Reply:
@jaideepr97 Charlie and I are going to look at it today and see how much of a lift it will be - if we can get it done without too much headache, we'll do it as part of this PR

Reply:
It would be ideal if you could follow the split structure, but if it's too much of a lift for this PR, then JD and I will take it on

Reply:
Sounds good - much like the `model` flag conversation, we really need to surface these conventions somewhere more obvious - otherwise we're going to discourage upstream contributions.

Reply:
To close the loop here, I will indeed be doing the splitting as part of this PR :smile: